eppoclient/evalbandits.go

```diff
@@ -0,0 +1,234 @@
+		case int64:
+			result.Numeric[key] = float64(value)
+		case uint:
+			result.Numeric[key] = float64(value)
+		case uint8:
+			result.Numeric[key] = float64(value)
+		case uint16:
+			result.Numeric[key] = float64(value)
+		case uint32:
+			result.Numeric[key] = float64(value)
+		case uint64:
+			result.Numeric[key] = float64(value)
+		case float32:
+			result.Numeric[key] = float64(value)
+		case float64:
+			result.Numeric[key] = value
+		case string:
+			result.Categorical[key] = value
+		case bool:
+			result.Categorical[key] = strconv.FormatBool(value)
+		}
+	}
+	return result
+}
+
+func (self ContextAttributes) toGenericAttributes() Attributes {
+	result := make(Attributes)
+	for key, value := range self.Numeric {
+		result[key] = value
+	}
+	for key, value := range self.Categorical {
+		result[key] = value
+	}
+	return result
+}
+
+type banditEvaluationContext struct {
+	flagKey           string
+	subjectKey        string
+	subjectAttributes ContextAttributes
+	actions           map[string]ContextAttributes
+}
+
+type banditEvaluationDetails struct {
+	flagKey           string
+	subjectKey        string
+	subjectAttributes ContextAttributes
+	result            *banditEvaluationResult
+}
+type banditEvaluationResult struct {
+	actionKey        string
+	actionAttributes ContextAttributes
+	actionScore      float64
+	actionWeight     float64
+	gamma            float64
+	optimalityGap    float64
+}
+
+type action struct {
+	key        string
+	attributes ContextAttributes
+}
+
+func (model *banditModelData) evaluate(ctx banditEvaluationContext) banditEvaluationDetails {
+	// THere's currently no way to change totalShards in bandit evaluation.
+	var totalShards int64 = 10_000
+
+	nActions := len(ctx.actions)
+	if nActions == 0 {
+		return banditEvaluationDetails{
+			flagKey:           ctx.flagKey,
+			subjectKey:        ctx.subjectKey,
+			subjectAttributes: ctx.subjectAttributes,
+			result:            nil,
+		}
+	}
+
+	bestAction, bestScore := "", math.Inf(-1)
+	scores := make(map[string]float64)
+	for actionKey, actionAttributes := range ctx.actions {
+		score := model.scoreAction(ctx.subjectAttributes, action{key: actionKey, attributes: actionAttributes})
+		scores[actionKey] = score
+		if score > bestScore {
+			bestAction, bestScore = actionKey, score
+		}
+	}
+
+	// adjust probability floor for number of actions to control the sum
+	minProbability := model.ActionProbabilityFloor / float64(nActions)
+
+	weights := make(map[string]float64)
+	totalWeight := 0.0
+	for actionKey, score := range scores {
+		if actionKey != bestAction { // best action is assigned the remainder weight
+			weight := math.Max(minProbability, 1.0/(float64(nActions)+model.Gamma*(bestScore-score)))
+			weights[actionKey] = weight
+			totalWeight += weight
+		}
+	}
+
+	remainderWeight := math.Max(0.0, 1.0-totalWeight)
+	weights[bestAction] = remainderWeight
+
+	actions := make([]string, nActions)
+	shards := make(map[string]int64)
```

Comment:
Consider a comment explaining what we are doing here: deterministically randomly shuffling the shards, so that as weights grow and shrink a huge tranch of users is not shifted from one old action to all now getting the same new action